File: backend/app/services/feature_extractor.py
"""
Feature Extractor — Production-grade URL & Text feature extraction.
Extracts 20+ URL features matching the ML model's expected input vector.
"""
import math
import re
from urllib.parse import urlparse, parse_qs, unquote


# ─── Constants ───────────────────────────────────────────────────────────────

URL_SHORTENERS = {
    "bit.ly", "tinyurl.com", "goo.gl", "t.co", "is.gd", "ow.ly",
    "buff.ly", "adf.ly", "bitly.com", "rb.gy", "cutt.ly", "short.io",
    "snip.ly", "bl.ink", "rebrand.ly", "su.pr", "fur.ly", "mcaf.ee",
    "links.co", "s.id", "n9.cl", "clck.ru", "clickmeter.com",
}

SUSPICIOUS_KEYWORDS = [
    "login", "signin", "sign-in", "logon", "verify", "verification",
    "update", "secure", "security", "account", "banking", "auth",
    "authentication", "confirm", "confirmation", "password", "passwd",
    "credential", "free", "gift", "paypal", "apple", "microsoft",
    "google", "amazon", "facebook", "netflix", "wallet", "bitcoin",
    "crypto", "recover", "reset", "unlock", "suspended", "alert",
    "urgent", "expire", "click-here", "click_here", "limited",
    "exclusive", "claim", "prize", "winner", "payment", "otp", "kyc",
]

BANKING_KEYWORDS = [
    "bank", "banking", "netbanking", "onlinebank", "bankofamerica",
    "wellsfargo", "citi", "chase", "barclays", "hsbc", "sbi",
    "icici", "hdfc", "axis", "ubi", "rbl", "kotak", "indusind",
]

SUSPICIOUS_TLDS = {
    "xyz", "top", "tk", "ml", "ga", "cf", "gq", "info", "click",
    "link", "online", "site", "website", "win", "download", "stream",
    "review", "loan", "science", "work", "party", "trade", "date",
    "racing", "bid", "webcam", "accountant", "faith", "cricket",
    "men", "gdn", "kim", "country", "pw", "zip",
}

TOP_BRANDS = [
    "paypal", "apple", "microsoft", "google", "amazon", "facebook",
    "instagram", "twitter", "linkedin", "netflix", "bankofamerica",
    "wellsfargo", "citi", "chase", "hsbc", "sbi", "icici",
    "dropbox", "icloud", "outlook", "yahoo", "ebay", "whatsapp",
]


# ─── Entropy ─────────────────────────────────────────────────────────────────

def calculate_entropy(text: str) -> float:
    """Shannon Entropy: H(X) = -Σ P(x) * log2(P(x))"""
    if not text:
        return 0.0
    freq = {}
    for c in text:
        freq[c] = freq.get(c, 0) + 1
    length = len(text)
    return round(-sum((f / length) * math.log2(f / length) for f in freq.values()), 4)


# ─── URL Feature Extraction ───────────────────────────────────────────────────

def extract_url_features(url: str) -> dict:
    """
    Extracts 20+ production-grade URL features.
    The first 7 features match the ML model's expected input vector exactly:
      [urlLength, domainLength, subdomains, hasIpAddress, isHttps, entropy, suspiciousKeywords]
    """
    # Normalize URL
    if "://" not in url:
        url = "http://" + url

    try:
        parsed = urlparse(url)
    except Exception:
        return _empty_url_features()

    full_domain = parsed.netloc or ""
    domain = full_domain.split(":")[0].lower()  # Strip port
    path = parsed.path or ""
    query_string = parsed.query or ""
    fragment = parsed.fragment or ""

    # Domain parts
    domain_parts = domain.split(".")
    tld = domain_parts[-1] if len(domain_parts) > 1 else ""
    # Subdomains = number of dots in host (e.g. "a.b.c.com" → 3 dots, 2 subdomains)
    subdomain_count = max(0, len(domain_parts) - 2)

    # IP detection
    ip_pattern = re.compile(r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$")
    has_ip = bool(ip_pattern.match(domain))

    # HTTPS
    is_https = parsed.scheme == "https"

    # Keyword detection (only on path + query, not domain name itself if it IS the brand)
    # This prevents google.com from self-flagging "google" as a keyword
    root_domain = ".".join(domain_parts[-2:]) if len(domain_parts) >= 2 else domain
    path_and_query = (path + "?" + query_string).lower()

    # Detect keywords in path+query for known domains, full URL for unknown ones
    keyword_check_target = url.lower()
    # If the URL is a major legitimate brand domain, only check path/query
    known_safe_roots = {
        "google.com", "microsoft.com", "apple.com", "amazon.com",
        "facebook.com", "instagram.com", "twitter.com", "linkedin.com",
        "netflix.com", "paypal.com", "dropbox.com", "outlook.com",
        "yahoo.com", "ebay.com", "github.com", "stackoverflow.com",
    }
    if root_domain in known_safe_roots:
        keyword_check_target = path_and_query

    keyword_matches = [kw for kw in SUSPICIOUS_KEYWORDS if kw in keyword_check_target]
    banking_matches = [kw for kw in BANKING_KEYWORDS if kw in keyword_check_target]

    # Query parameters
    qs_params = parse_qs(query_string)

    # Encoded characters
    encoded_count = url.count("%")

    # Port
    port = parsed.port  # None if default

    # Special chars in domain
    special_chars_domain = len(re.findall(r"[^a-zA-Z0-9.\-]", domain))

    # Path depth
    path_depth = len([p for p in path.split("/") if p])

    # Hyphens and digits in domain
    hyphens_in_domain = domain.count("-")
    digits_in_domain = sum(c.isdigit() for c in domain)

    # Check for @ symbol (redirect obfuscation)
    has_at_sign = "@" in url

    # Check for double slash after path start
    has_double_slash = "//" in path

    return {
        # ── Core 7 Features (ML model input vector) ──
        "urlLength": len(url),
        "domainLength": len(domain),
        "subdomains": subdomain_count,
        "hasIpAddress": has_ip,
        "isHttps": is_https,
        "entropy": calculate_entropy(url),
        "suspiciousKeywords": len(keyword_matches),

        # ── Extended Features ──
        "tld": tld,
        "hyphensInDomain": hyphens_in_domain,
        "digitsInDomain": digits_in_domain,
        "queryParams": len(qs_params),
        "encodedCharacters": encoded_count,
        "pathDepth": path_depth,
        "hasAtSign": has_at_sign,
        "hasDoubleSlash": has_double_slash,
        "hasFragment": bool(fragment),
        "hasPort": port is not None,
        "port": port or 0,
        "isShortener": domain in URL_SHORTENERS,
        "bankingKeywords": len(banking_matches),
        "specialCharsInDomain": special_chars_domain,
        "domainParts": len(domain_parts),
        "isHttpsFlag": 1 if is_https else 0,
        "keywordMatches": keyword_matches[:10],  # Cap for JSON size
    }


def extract_text_features(text: str) -> dict:
    """Extracts features from free-text payloads (Email, SMS, WhatsApp, Image OCR)."""
    text_lower = text.lower()

    suspicious_kws = [
        "urgent", "verify", "otp", "password", "bank", "account",
        "locked", "suspended", "prize", "winner", "kyc", "investment",
        "crypto", "delivery", "package", "click", "confirm", "login",
        "update", "secure", "alert", "expire", "unusual", "unauthorized",
    ]
    keyword_matches = [kw for kw in suspicious_kws if kw in text_lower]

    urls_in_text = re.findall(r"(https?://[^\s<>\"']+)", text)
    phone_numbers = re.findall(r"\+?\d[\d\s\-]{9,14}\d", text)

    return {
        "textLength": len(text),
        "suspiciousKeywords": len(keyword_matches),
        "urlsCount": len(urls_in_text),
        "urls": urls_in_text[:10],
        "phoneNumbersCount": len(phone_numbers),
        "entropy": calculate_entropy(text),
    }


def _empty_url_features() -> dict:
    return {
        "urlLength": 0, "domainLength": 0, "subdomains": 0,
        "hasIpAddress": False, "isHttps": False, "entropy": 0.0,
        "suspiciousKeywords": 0, "tld": "", "hyphensInDomain": 0,
        "digitsInDomain": 0, "queryParams": 0, "encodedCharacters": 0,
        "pathDepth": 0, "hasAtSign": False, "hasDoubleSlash": False,
        "hasFragment": False, "hasPort": False, "port": 0,
        "isShortener": False, "bankingKeywords": 0,
        "specialCharsInDomain": 0, "domainParts": 0, "isHttpsFlag": 0,
        "keywordMatches": [],
    }

File: frontend/src/lib/engine/featureExtractor.ts
export const URL_SHORTENERS = new Set([
  "bit.ly", "tinyurl.com", "goo.gl", "t.co", "is.gd", "ow.ly",
  "buff.ly", "adf.ly", "bitly.com", "rb.gy", "cutt.ly", "short.io",
  "snip.ly", "bl.ink", "rebrand.ly", "su.pr", "fur.ly", "mcaf.ee",
  "links.co", "s.id", "n9.cl", "clck.ru", "clickmeter.com",
]);

export const SUSPICIOUS_TLDS = new Set([
  "tk", "ml", "ga", "cf", "gq", "icu", "top", "xyz", "club", "site",
  "vip", "wang", "win", "bid", "stream", "men", "party", "pro", "link",
  "review", "country", "kim", "science", "work", "asia", "rest", "surf"
]);

export const SUSPICIOUS_KEYWORDS = [
  "login", "signin", "sign-in", "logon", "verify", "verification",
  "update", "secure", "security", "account", "banking", "auth",
  "authentication", "confirm", "confirmation", "password", "passwd",
  "credential", "free", "gift", "paypal", "apple", "microsoft",
  "google", "amazon", "facebook", "netflix", "wallet", "bitcoin",
  "crypto", "recover", "reset", "unlock", "suspended", "alert",
  "urgent", "expire", "click-here", "click_here", "limited",
  "exclusive", "claim", "prize", "winner", "payment", "otp", "kyc",
];

export const BANKING_KEYWORDS = [
  "bank", "banking", "netbanking", "onlinebank", "bankofamerica",
  "wellsfargo", "citi", "chase", "barclays", "hsbc", "sbi",
  "icici", "hdfc", "axis", "ubi", "rbl", "kotak", "indusind",
];

export function calculateEntropy(text: string): number {
  if (!text) return 0.0;
  const freq: Record<string, number> = {};
  for (let i = 0; i < text.length; i++) {
    const c = text[i];
    freq[c] = (freq[c] || 0) + 1;
  }
  let entropy = 0;
  for (const f of Object.values(freq)) {
    const p = f / text.length;
    entropy -= p * Math.log2(p);
  }
  return parseFloat(entropy.toFixed(4));
}

export function extractUrlFeatures(url: string) {
  let normalizedUrl = url.trim();
  if (!normalizedUrl.includes("://")) {
    normalizedUrl = "http://" + normalizedUrl;
  }

  let parsed: URL;
  try {
    parsed = new URL(normalizedUrl);
  } catch {
    return _emptyUrlFeatures();
  }

  const fullDomain = parsed.hostname;
  const domain = fullDomain.toLowerCase();
  const path = parsed.pathname;
  const queryString = parsed.search;
  const fragment = parsed.hash;

  const domainParts = domain.split(".");
  const tld = domainParts.length > 1 ? domainParts[domainParts.length - 1] : "";
  const subdomains = Math.max(0, domainParts.length - 2);

  const ipPattern = /^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/;
  const hasIp = ipPattern.test(domain);

  const isHttps = parsed.protocol === "https:";

  const rootDomain = domainParts.length >= 2 ? domainParts.slice(-2).join(".") : domain;
  const pathAndQuery = (path + queryString).toLowerCase();

  const knownSafeRoots = new Set([
    "google.com", "microsoft.com", "apple.com", "amazon.com",
    "facebook.com", "instagram.com", "twitter.com", "linkedin.com",
    "netflix.com", "paypal.com", "dropbox.com", "outlook.com",
    "yahoo.com", "ebay.com", "github.com", "stackoverflow.com",
  ]);

  const keywordCheckTarget = knownSafeRoots.has(rootDomain) ? pathAndQuery : url.toLowerCase();

  const keywordMatches = SUSPICIOUS_KEYWORDS.filter((kw) => keywordCheckTarget.includes(kw));
  const bankingMatches = BANKING_KEYWORDS.filter((kw) => keywordCheckTarget.includes(kw));

  const queryParams = Array.from(parsed.searchParams.keys()).length;
  const encodedCount = (url.match(/%/g) || []).length;
  const port = parsed.port ? parseInt(parsed.port, 10) : 0;
  const specialCharsDomain = (domain.match(/[^a-zA-Z0-9.\-]/g) || []).length;
  const pathDepth = path.split("/").filter((p) => p).length;
  const hyphensInDomain = (domain.match(/-/g) || []).length;
  const digitsInDomain = (domain.match(/\d/g) || []).length;
  const hasAtSign = url.includes("@");
  const hasDoubleSlash = path.includes("//");

  return {
    urlLength: url.length,
    domainLength: domain.length,
    subdomains: subdomains,
    hasIpAddress: hasIp,
    isHttps: isHttps,
    entropy: calculateEntropy(url),
    suspiciousKeywords: keywordMatches.length,
    tld: tld,
    hyphensInDomain: hyphensInDomain,
    digitsInDomain: digitsInDomain,
    queryParams: queryParams,
    encodedCharacters: encodedCount,
    pathDepth: pathDepth,
    hasAtSign: hasAtSign,
    hasDoubleSlash: hasDoubleSlash,
    hasFragment: !!fragment,
    hasPort: !!parsed.port,
    port: port,
    isShortener: URL_SHORTENERS.has(domain),
    bankingKeywords: bankingMatches.length,
    specialCharsInDomain: specialCharsDomain,
    domainParts: domainParts.length,
    isHttpsFlag: isHttps ? 1 : 0,
    keywordMatches: keywordMatches.slice(0, 10),
  };
}

export function extractTextFeatures(text: string) {
  const textLower = text.toLowerCase();
  const suspiciousKws = [
    "urgent", "verify", "otp", "password", "bank", "account",
    "locked", "suspended", "prize", "winner", "kyc", "investment",
    "crypto", "delivery", "package", "click", "confirm", "login",
    "update", "secure", "alert", "expire", "unusual", "unauthorized",
  ];
  const keywordMatches = suspiciousKws.filter((kw) => textLower.includes(kw));
  const urlsInText = text.match(/(https?:\/\/[^\s<>"']+)/g) || [];
  const phoneNumbers = text.match(/\+?\d[\d\s\-]{9,14}\d/g) || [];

  return {
    textLength: text.length,
    suspiciousKeywords: keywordMatches.length,
    urlsCount: urlsInText.length,
    urls: urlsInText.slice(0, 10),
    phoneNumbersCount: phoneNumbers.length,
    entropy: calculateEntropy(text),
  };
}

function _emptyUrlFeatures() {
  return {
    urlLength: 0, domainLength: 0, subdomains: 0,
    hasIpAddress: false, isHttps: false, entropy: 0.0,
    suspiciousKeywords: 0, tld: "", hyphensInDomain: 0,
    digitsInDomain: 0, queryParams: 0, encodedCharacters: 0,
    pathDepth: 0, hasAtSign: false, hasDoubleSlash: false,
    hasFragment: false, hasPort: false, port: 0,
    isShortener: false, bankingKeywords: 0,
    specialCharsInDomain: 0, domainParts: 0, isHttpsFlag: 0,
    keywordMatches: [],
  };
}
